### PR TITLE
[Merged by Bors] - feat: arctangent addition and Machin's formula for π

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
`four_mul_arctan_inv_5_sub_arctan_inv_239` is theorem 1 in my own #6091.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean [hunk 1]
```diff
@@ -927,6 +927,14 @@ theorem tan_pi_div_four : tan (π / 4) = 1 := by
 theorem tan_pi_div_two : tan (π / 2) = 0 := by simp [tan_eq_sin_div_cos]
 #align real.tan_pi_div_two Real.tan_pi_div_two
 
+@[simp]
+theorem tan_pi_div_six : tan (π / 6) = 1 / sqrt 3 := by
+  rw [tan_eq_sin_div_cos, sin_pi_div_six, cos_pi_div_six]; ring
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean:932]:
```suggestion
  rw [tan_eq_sin_div_cos, sin_pi_div_six, cos_pi_div_six]
  ring

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean [hunk 2]
```diff
@@ -927,6 +927,14 @@ theorem tan_pi_div_four : tan (π / 4) = 1 := by
...
+
+@[simp]
+theorem tan_pi_div_three : tan (π / 3) = sqrt 3 := by
+  rw [tan_eq_sin_div_cos, sin_pi_div_three, cos_pi_div_three]; ring
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean:936]: Same here.

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean [hunk 3]
```diff
@@ -200,6 +200,83 @@ theorem arccos_eq_arctan {x : ℝ} (h : 0 < x) : arccos x = arctan (sqrt (1 - x
   · linarith only [arcsin_pos.2 h]
 #align real.arccos_eq_arctan Real.arccos_eq_arctan
 
+theorem arctan_inv_of_pos {x : ℝ} (h : 0 < x) : arctan x⁻¹ = π / 2 - arctan x := by
+  rw [← arctan_tan (x := _ - _), tan_pi_div_two_sub, tan_arctan]
+  · norm_num; exact (arctan_lt_pi_div_two x).trans (half_lt_self_iff.mpr pi_pos)
+  · rw [sub_lt_self_iff, ← arctan_zero]; exact tanOrderIso.symm.strictMono h
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:206]: My reading of the style guide is that putting each tactic call on a single line is slightly preferred. Can you do this (also in the other places)?

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean [hunk 4]
```diff
@@ -200,6 +200,83 @@ theorem arccos_eq_arctan {x : ℝ} (h : 0 < x) : arccos x = arctan (sqrt (1 - x
...
+    (h₂ : arctan x + arctan y < π / 2) : arctan x + arctan y = arctan ((x + y) / (1 - x * y)) := by
+  rw [← arctan_tan h₁ h₂]; congr
+  conv_rhs => rw [← tan_arctan x, ← tan_arctan y]
+  exact tan_add' ⟨arctan_ne_mul_pi_div_two, arctan_ne_mul_pi_div_two⟩
+
+lemma arctan_add_arctan_lt_pi_div_two {x y : ℝ} (h : x * y < 1) : arctan x + arctan y < π / 2 := by
+  cases' le_or_lt y 0 with hy hy
+  · rw [← add_zero (π / 2), ← arctan_zero]
+    exact add_lt_add_of_lt_of_le (arctan_lt_pi_div_two _) (tanOrderIso.symm.monotone hy)
+  rw [← lt_div_iff hy, ← inv_eq_one_div] at h
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:232]: Can you indent these subgoals with a focusing dot also?

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean [hunk 5]
```diff
@@ -200,6 +200,83 @@ theorem arccos_eq_arctan {x : ℝ} (h : 0 < x) : arccos x = arctan (sqrt (1 - x
...
+
+theorem two_mul_arctan_add_pi {x : ℝ} (h : 1 < x) :
+    2 * arctan x = arctan (2 * x / (1 - x ^ 2)) + π := by
+  rw [two_mul, arctan_add_eq_add_pi (by nlinarith) (by linarith)]; congr 2; ring
+
+theorem two_mul_arctan_sub_pi {x : ℝ} (h : x < -1) :
+    2 * arctan x = arctan (2 * x / (1 - x ^ 2)) - π := by
+  rw [two_mul, arctan_add_eq_sub_pi (by nlinarith) (by linarith)]; congr 2; ring
+
+/-- **Machin's formula for π**. -/
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:274]: Is there only one such formula? Would putting this in the docstring be helpful? (I wouldn't have recognised the name, for instance.)

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean [hunk 6]
```diff
@@ -200,6 +200,83 @@ theorem arccos_eq_arctan {x : ℝ} (h : 0 < x) : arccos x = arctan (sqrt (1 - x
   · linarith only [arcsin_pos.2 h]
 #align real.arccos_eq_arctan Real.arccos_eq_arctan
 
+theorem arctan_inv_of_pos {x : ℝ} (h : 0 < x) : arctan x⁻¹ = π / 2 - arctan x := by
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:203]: Can to put these lemmas into a separate section, as appropriate, and mention them in the module docstring?

**Review by grunweg — COMMENTED** [review]
I took a look and left a few comments, mostly about style. (Some are certainly subjective.)

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean (hunk 3 above)
Comment (Parcly-Taxel) [reply Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:206]: Existing code in `Arctan.lean` already contains some instances of multiple tactics on the same line (e.g. `arctan_eq_arccos`). I'm keeping my own instances of this in `Arctan.lean` but not in `Basic.lean`.

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean (hunk 4 above)
Comment (Parcly-Taxel) [reply Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:232]: Done.

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean (hunk 6 above)
Comment (Parcly-Taxel) [reply Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:209]: Done.

**Comment by jcommelin** [comment]
LGTM, but someone more familiar with this part of the library might want to take another look.

maintainer merge

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean [hunk 7]
```diff
@@ -200,6 +206,101 @@ theorem arccos_eq_arctan {x : ℝ} (h : 0 < x) : arccos x = arctan (sqrt (1 - x
...
+  · rw [← add_zero (π / 2), ← arctan_zero]
+    exact add_lt_add_of_lt_of_le (arctan_lt_pi_div_two _) (tanOrderIso.symm.monotone hy)
+  · rw [← lt_div_iff hy, ← inv_eq_one_div] at h
+    replace h : arctan x < arctan y⁻¹ := tanOrderIso.symm.strictMono h
+    rwa [arctan_inv_of_pos hy, lt_tsub_iff_right] at h
+
+theorem arctan_add {x y : ℝ} (h : x * y < 1) :
+    arctan x + arctan y = arctan ((x + y) / (1 - x * y)) := by
+  apply arctan_add'
+  rw [neg_lt, neg_add, ← arctan_neg, ← arctan_neg]
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:249]: At this point, you have two goals. This should show up in the proof structure, with focusing dots.

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean [hunk 8]
```diff
@@ -200,6 +206,101 @@ theorem arccos_eq_arctan {x : ℝ} (h : 0 < x) : arccos x = arctan (sqrt (1 - x
...
+
+theorem arctan_add_eq_sub_pi {x y : ℝ} (h : 1 < x * y) (hx : x < 0) :
+    arctan x + arctan y = arctan ((x + y) / (1 - x * y)) - π := by
+  rw [← neg_mul_neg] at h
+  have k := arctan_add_eq_add_pi h (neg_pos.mpr hx)
+  rw [show _ / _ = -((x + y) / (1 - x * y)) by ring, ← neg_inj] at k
+  simp only [arctan_neg, neg_add, neg_neg, ← sub_eq_add_neg _ π] at k
+  exact k
+
+theorem two_mul_arctan {x : ℝ} (h : x * x < 1) : 2 * arctan x = arctan (2 * x / (1 - x ^ 2)) := by
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:273]: The assumption `x ^ 2 < 1` would be more standard, or even better `-1 < x` and `x < 1`.

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean [hunk 9]
```diff
@@ -200,6 +206,101 @@ theorem arccos_eq_arctan {x : ℝ} (h : 0 < x) : arccos x = arctan (sqrt (1 - x
...
+    2 * arctan x = arctan (2 * x / (1 - x ^ 2)) + π := by
+  rw [two_mul, arctan_add_eq_add_pi (by nlinarith) (by linarith)]; congr 2; ring
+
+theorem two_mul_arctan_sub_pi {x : ℝ} (h : x < -1) :
+    2 * arctan x = arctan (2 * x / (1 - x ^ 2)) - π := by
+  rw [two_mul, arctan_add_eq_sub_pi (by nlinarith) (by linarith)]; congr 2; ring
+
+theorem arctan_inv_2_add_arctan_inv_3 : arctan 2⁻¹ + arctan 3⁻¹ = π / 4 := by
+  rw [arctan_add]
+  all_goals norm_num
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean:286]:
```suggestion
  rw [arctan_add] <;> norm_num
```
Same thing below.

**Review by sgouezel — COMMENTED** [review]
Nice, thanks!

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean (hunk 9 above)
Comment (Parcly-Taxel) [reply Mathlib/Analysis/SpecialFunctions/Trigonometric/Arctan.lean]: All done.

**Pull request closed without merging**